### PR TITLE
BUG: DateTimeIndex.is_year_start unexpected behavior when constructed with freq 'MS' date_range (#57377)

### LINKED ISSUE #49606: BUG: `DatetimeIndex.is_year_start` returns incorrect array if `.freq == 'MS'`
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd

i = pd.date_range('2020-01-01', freq='MS', periods=12)
# DatetimeIndex(['2020-01-01', '2020-02-01', '2020-03-01', '2020-04-01',
#                '2020-05-01', '2020-06-01', '2020-07-01', '2020-08-01',
#                '2020-09-01', '2020-10-01', '2020-11-01', '2020-12-01'],
#               dtype='datetime64[ns]', freq='MS')

# Incorrect result:
i.is_year_start 
# array([False, False, False, False, False, False, False, False, False, False, False, True])

# Correct result when first removing the frequency:
i.freq = None
i.is_year_start
# array([True, False, False, False, False, False, False, False, False, False, False, False])
```

### Issue Description

The `.freq` property of the `DatetimeIndex` influences the result of the `.is_year_start` property. If it is set to MonthStart, the december timestamp is incorrectly recognised as the start of the year.

### Expected Behavior

Even if `i.freq == 'MS'`, the result of `i.is_year_start` should start with `True`, not end with `True`.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 91111fd99898d9dcaa6bf6bedb662db4108da6e6
python           : 3.9.13.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19045
machine          : AMD64
processor        : Intel64 Family 6 Model 165 Stepping 2, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : de_DE.cp1252

pandas           : 1.5.1
numpy            : 1.23.4
pytz             : 2022.6
dateutil         : 2.8.2
setuptools       : 63.4.1
pip              : 22.2.2
Cython           : None
pytest           : 7.1.3
hypothesis       : None
sphinx           : 5.3.0
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : 8.5.0
pandas_datareader: None
bs4              : 4.11.1
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.6.2
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : 3.0.10
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None
tzdata           : None
</details>

### LINKED ISSUE #57377: BUG: Inconsistent "is_year_start" from DatetimeIndex with freq "MS"
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd
dr = pd.date_range("2017-12-01", periods=2, freq="MS")
dr_comp = [dt.is_year_start for dt in dr]
dr_attr = dr.is_year_start
assert dr[1].is_year_start
assert dr_comp[1]
assert dr_attr[0]  # Should raise error but doesn't
assert dr_attr[1]  # Raises error but shouldn't
```

### Issue Description

When working with DatetimeIndex with frequency "MS" the attribute "is_year_start" does not obtain the same attribute if you obtain the attribute directly from the DatatimeIndex instead of obtaining directly from each individual value of the DatetimeIndex.

### Expected Behavior

I would expect for the same array of attributes independently of if it is obtained from a comprehensive list or as an attibute as shown in the documentation.

https://pandas.pydata.org/docs/reference/api/pandas.DatetimeIndex.is_year_start.html

```python
import pandas as pd
dr = pd.date_range("2017-12-01", periods=2, freq="MS")
dr_comp = [dt.is_year_start for dt in dr]
dr_attr = dr.is_year_start
assert not dr[1].is_year_start
assert dr[1].is_year_start
assert not dr_comp[1]
assert dr_comp[1]
assert not dr_attr[0] 
assert dr_attr[1] 
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : f538741432edf55c6b9fb5d0d496d2dd1d7c2457
python                : 3.11.6.final.0
python-bits           : 64
OS                    : Linux
OS-release            : 3.10.0-1160.95.1.el7.x86_64
Version               : #1 SMP Fri Jun 23 08:44:55 EDT 2023
machine               : x86_64
processor             : x86_64
byteorder             : little
LC_ALL                : en_US.UTF-8
LANG                  : en_US.UTF-8
LOCALE                : en_US.UTF-8

pandas                : 2.2.0
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.8.2
setuptools            : 68.2.2
pip                   : 23.2.1
Cython                : None
pytest                : None
hypothesis            : None
sphinx                : None
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : None
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : None
IPython               : 8.16.1
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : None
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : None
gcsfs                 : None
matplotlib            : None
numba                 : None
numexpr               : None
odfpy                 : None
openpyxl              : 3.1.2
pandas_gbq            : None
pyarrow               : None
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : None
sqlalchemy            : 2.0.21
tables                : None
tabulate              : None
xarray                : None
xlrd                  : None
zstandard             : None
tzdata                : 2023.4
qtpy                  : None
pyqt5                 : None

</details>

### PR DESCRIPTION
**Pull request by mattheeter**
- [x] closes #49606
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

- A DateTimeIndex created via date_range with the freq parameter as 'MS' would yield a wrong is_year_start (and later it was discovered is_quarter_start) accessor. In `pandas/_libs/tslibs/fields.pyx`, the way in which the start and end months were set was one way for freq 'MS', 'QS', and 'YS' and another way for other offsets. However, only the latter two had the necessary properties to set start and end this way. As such, removing 'MS' from the list fixed the issue.

### CODE REVIEW

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Comment by MarcoGorelli** [comment]
wait sorry, I think this is my fault for not having got round to reviewing it

reopening for now, hoping to get to it at some point before 3.0

@natmokval fancy taking a look too?

**Comment by natmokval** [comment]
I think the suggested changes will fix the bug in `is_year_start` for `MonthBegin`. We will get the correct result for `freq="MS"`:
```
dr = pd.date_range("2017-12-01", periods=2, freq="MS")
print(dr) # DatetimeIndex(['2017-12-01', '2018-01-01'], dtype='datetime64[ns]', freq='MS')
dr_attr = dr.is_year_start
print(dr_attr) # [False  True]
```

These changes won't fix a problem with some other frequencies such as: `freq="QS-NOV"`:
```
dr = pd.date_range("2017-07-01", periods=2, freq="QS-NOV")
print(dr) # DatetimeIndex(['2017-08-01', '2017-11-01'], dtype='datetime64[ns]', freq='QS-NOV')
dr_comp = [dt.is_year_start for dt in dr]
dr_attr = dr.is_year_start
print(dr_attr) # [False  True]
print(dr_comp) # [False, False]
```

**Comment by natmokval** [comment]
I think it's okay to leave the fix for `"MS"` in this PR and open a new PR to work on the problem for frequencies such as `"QS-NOV"`.

This PR is still `Stale`. It can take a couple of days to change its status to active.

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
thanks for your PR - I think a more general fix is needed here, rather than papering over already-broken logic

EDIT: I've made a separate issue for the other problem: https://github.com/pandas-dev/pandas/issues/58523

File: pandas/_libs/tslibs/fields.pyx [hunk 1]
```diff
@@ -253,10 +253,12 @@ def get_start_end_field(
         # month of year. Other offsets use month, startingMonth as ending
         # month of year.
 
-        if (freqstr[0:2] in ["MS", "QS", "YS"]) or (
-                freqstr[1:3] in ["MS", "QS", "YS"]):
+        # According to the above comment, the first conditional is for QS and YS only
+        if (freqstr[0:2] in ["QS", "YS"]) or (
+                freqstr[1:3] in ["QS", "YS"]):
```
[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply pandas/_libs/tslibs/fields.pyx:258]: wow, looks like this broken logic was introduced 13 years ago... 😭 

https://github.com/pandas-dev/pandas/pull/4823/files

**Review by MarcoGorelli — APPROVED** [review]
Thanks @mattheeter for your PR, thanks for your patience here, and sorry that this took a while

Looks good to me. I've added a hypothesis test too to make sure

**Pull request merged**